### PR TITLE
Add boxer

### PR DESCRIPTION
**Pull request by mixmix**
In working on private groups I'm keen to register new boxers (as well as unboxers).

This PR adds:
- adds `db.addBoxer` which follows a similar pattern to `addUnboxer`
- moves private-box (box1) code to use the new API (in preparation to extract it!)

### CODE REVIEW

**Comment by staltz** [comment]
I like what I see but I haven't reviewed it yet, and being about privacy and touching ssb-db, this needs special attention. Will (probably) come back to this and review properly.

File: index.js [hunk 1]
```diff
@@ -106,25 +106,28 @@ module.exports = {
...
+        return ssb[name]
+      },
 
       close                    : close,
-      del: valid.async(ssb.del, 'msgLink|feedId'),
+      del                      : valid.async(ssb.del, 'msgLink|feedId'),
       publish                  : valid.async(feed.add, 'string|msgContent'),
       add                      : valid.async(ssb.add, 'msg'),
-      queue                      : valid.async(ssb.queue, 'msg'),
+      queue                    : valid.async(ssb.queue, 'msg'),
```
Comment (mixmix) [inline index.js]: sorry, these misaligned dots have been eating my brain for 1-2 years.
I couldn't leave them alone

File: create.js [hunk 2]
```diff
@@ -2,11 +2,12 @@
 
 var join = require('path').join
 var EventEmitter = require('events')
-
 var pull = require('pull-stream')
 var ref = require('ssb-ref')
 var ssbKeys = require('ssb-keys')
 
+var createDB = require('./db')
+var extras = require('./extras')
```
Comment (mixmix) [inline create.js]: I pulled these to the top of the file because I like to see what's been pulled into a file
I was passing through here trying to understand where the addUnboxer connected in an how.
I hope this help later people reading to read more easily

File: test/end-to-end.js [hunk 3]
```diff
@@ -99,11 +95,11 @@ module.exports = function (opts) {
...
           // bob can also decrypt
           var content = ssbKeys.unbox(ctxt, alice.private)
-          t.deepEqual(content, {
-            type: 'secret2',
-            secret: "it's a secret!",
-            recps: [alice.id]
-          }, 'alice can decrypt')
+          t.deepEqual(
+            content,
+            { type: 'secret2', secret: "it's a secret!", recps: [alice.id] },
```
Comment (mixmix) [inline test/end-to-end.js]: :fire: this test does something very strange (I think).
if you compare what was published with `feed.add` and what comes out the the private box ... they're different!

The RECPS field has been array'd !
This is fine, so long as it doesn't mess with signature / hashes of content?
Maybe I don't understand the order of that relative to encryption though

Comment (mixmix) [reply test/end-to-end.js]: _I didn't change the test, just the layout to make it easy to compare. The code I wrote still passes this strange test_

File: create.js (hunk 2 above)
Comment (christianbundy) [reply create.js]: Very much appreciated, thank you.

File: minimal.js [hunk 4]
```diff
@@ -208,34 +221,48 @@ module.exports = function (dirname, keys, opts) {
...
-      } else {
-        throw new Error(
-          'private message recipients must be valid feeds, '+
-          'was:' + JSON.stringify(recps)
-        )
-      }
+  /* TODO extract to ssb-private */
+  var box1 = {
+    box: function (content, recps) {
+      if (!recps.every(isFeed)) return null
```
Comment (christianbundy) [inline minimal.js]: IIRC Patchbay sets `recps` as `{ link, name }`, is that something we want to support here?

File: minimal.js [hunk 5]
```diff
@@ -208,34 +221,48 @@ module.exports = function (dirname, keys, opts) {
...
+  db.addUnboxer(box1.unbox)
+  // ////////////////////////////////
+
+  // db.box = function (value, state) {
+  db.box = function (value) {
+    var recps = value.content.recps
+    if (!recps) return value.content
+
+    if (typeof recps === 'string') recps = value.content.recps = [recps]
+    if (!Array.isArray(recps)) throw new Error('private message field "recps" expects an Array of recipients')
```
Comment (christianbundy) [inline minimal.js]: Maybe redundant, but my intuition is that this change would throw errors when Patchbay makes private messages. If that's true, it would be important to ensure that this is published as a breaking change. FWIW, I'm a fan of the simplicity.

File: minimal.js [hunk 6]
```diff
@@ -208,34 +221,48 @@ module.exports = function (dirname, keys, opts) {
...
+  db.addBoxer(box1.box)
+  db.addUnboxer(box1.unbox)
+  // ////////////////////////////////
+
+  // db.box = function (value, state) {
+  db.box = function (value) {
+    var recps = value.content.recps
+    if (!recps) return value.content
+
+    if (typeof recps === 'string') recps = value.content.recps = [recps]
```
Comment (christianbundy) [inline minimal.js]: I'm comfortable with this, but would also be comfortable not supporting this use-case and just insisting that `recps` is an array.

File: minimal.js (hunk 4 above)
Comment (mixmix) [reply minimal.js]: I don't want to support everyone and their dog. I just want it to be an Array of `feedIds`. Keep in mind this is just the boxer, it doesn't change anything already published.

Also, I'm _fairly_ sure my code doesn't change any behaviour here...

File: minimal.js (hunk 6 above)
Comment (mixmix) [reply minimal.js]: yeah me too. I was trying to make this a PR added an API, but didn't change any behaviour...

File: minimal.js (hunk 4 above)
Comment (mixmix) [reply minimal.js]: Hey patchbay stuff was written before the magical `recps` auto-encryption (I think). Here's the code in patchcore which is doing the encryption for patchbay @christianbundy : https://github.com/ssbc/patchcore/blob/master/sbot.js#L163-L168

TL;DR - this isn't a breaking change

File: minimal.js (hunk 5 above)
Comment (mixmix) [reply minimal.js]: I don't see how this make a breaking change. See above comment

File: minimal.js (hunk 4 above)
Comment (christianbundy) [reply minimal.js]: Sweet! Thanks for learning me some history, I don't remember `recps` auto-encryption even being a new thing, but that's great info to have.

File: minimal.js (hunk 5 above)
Comment (christianbundy) [reply minimal.js]: Yup, my bad there.

File: create.js [hunk 7]
```diff
@@ -17,13 +18,13 @@ function errorCB (err) {
...
 
-module.exports = function (path, opts, keys) {
+module.exports = function create (path, opts, keys) {
   //_ was legacy db. removed that, but for backwards compatibilty reasons do not change interface
   if(!path) throw new Error('path must be provided')
 
   keys = keys || ssbKeys.generate()
 
-  var db = require('./db')(join(opts.path || path, 'flume'), keys, opts)
+  var db = createDB(join(opts.path || path, 'flume'), keys, opts)
```
Comment (arj03) [inline create.js]: This is much nicer to read ❤️

**Comment by mixmix** [comment]
@christianbundy @arj03 is this good enough to merge in your opinion.

Should I run other tests like symlink in ssb-server and see if breaks all those integrations?

**Comment by christianbundy** [comment]
The SSB-Server tests are consistenty broken for me every time I try them (including tonight). If you have a way of easily testing, that'd be cool, but I'd be comfortable just optimistically merging this.

**Comment by arj03** [comment]
I think if the tests passes for this repo it should be fine to merge.

**Comment by mixmix** [comment]
Published as `ssb-db@19.4.0`

**Comment by mixmix** [comment]
Thanks for help / input on this @arj03 + @christianbundy
